generator/template.go

```diff
@@ -195,6 +196,105 @@ function getNotifyEntityArrivalSink<T>(notifyCallback: NotifyStreamEntityArrival
     }
   })
 }
+
+type Primitive = string | boolean | number;
+type RequestPayload = Record<string, unknown>;
+type FlattenedRequestPayload = Record<string, Primitive | Array<Primitive>>;
+
+function isPlainObject(val: unknown): boolean {
+  const isObject =
+    Object.prototype.toString.call(val).slice(8, -1) === "Object";
+  const isObjLike = val !== null && isObject;
+
+  if (!isObjLike || !isObject) {
+    return false;
+  }
+
+  const proto = Object.getPrototypeOf(val);
+
+  const hasObjectConstructor =
+    typeof proto === "object" &&
+    proto.constructor === Object.prototype.constructor;
+
+  return hasObjectConstructor;
+}
+
+/**
+ * Checks if given value is of a primitive type
+ * @param  {unknown} value
+ * @return {boolean}
+ */
+function isPrimitive(value: unknown): boolean {
+  return ["string", "number", "boolean"].some(t => typeof value === t);
+}
+
+/**
+ * Flattens a deeply nested request payload and returns an object
+ * with only primitive values and non-empty array of primitive values
+ * as per https://github.com/googleapis/googleapis/blob/master/google/api/http.proto
+ * @param  {RequestPayload} requestPayload
+ * @param  {String} path
+ * @return {FlattenedRequestPayload>}
+ */
+function flattenRequestPayload<T extends RequestPayload>(
+  requestPayload: T,
+  path: string = ""
+): FlattenedRequestPayload {
+  return Object.keys(requestPayload).reduce(
+    (acc: T, key: string): T => {
+      const value = requestPayload[key];
+      const newPath = path ? [path, key].join(".") : key;
+
+      const isNonEmptyPrimitiveArray =
+        Array.isArray(value) &&
+        value.every(v => isPrimitive(v)) &&
+        value.length > 0;
+
+      let objectToMerge = {};
+
+      if (isPlainObject(value)) {
+        objectToMerge = flattenRequestPayload(value as RequestPayload, newPath);
+      } else if (isPrimitive(value) || isNonEmptyPrimitiveArray) {
```

Comment:
I think you misunderstood. the 0 & false logic here is more about. 

if say request object `{ a:0, c:false, b: 1)`, then the output query string should be `?b=1`, the current implementation doesn't cover it

Reply:
Hmmm... I think we should not filter out zero values from the query string as they may be intentionally added by the user. Let me know your opinion on this.

Reply:
it's not needed though. and jsonpb also ignores the default value to save space. we should follow the same.

https://developers.google.com/protocol-buffers/docs/proto3#json

Reply:
Ah I see.. Thanks for pointing this out... I will modify the flatten function to filter out keys with zero values 🙇🏼

Reply:
I have modified the flatten function to filter zero values as well 🙇🏼